esi/src/lib.rs

```diff
@@ -23,210 +22,395 @@
+                                original_request_metadata.clone_without_body(),
+                                &alt,
+                            )
+                        });
+
+                        if let Some(element) = send_fragment_request(
+                            req?,
+                            alt_req,
+                            continue_on_error,
+                            dispatch_fragment_request,
+                        )? {
+                            elements.push_back(element);
                         }
                     }
                     Event::XML(event) => {
-                        xml_writer.write_event(event)?;
-                        xml_writer.inner().flush().expect("failed to flush output");
+                        if elements.is_empty() {
+                            debug!("nothing waiting so streaming directly to client");
+                            output_writer.write_event(event)?;
+                            output_writer
+                                .inner()
+                                .flush()
+                                .expect("failed to flush output");
+                        } else {
+                            debug!("pushing content to buffer, len: {}", elements.len());
+                            let mut vec = Vec::new();
+                            let mut writer = Writer::new(&mut vec);
+                            writer.write_event(event)?;
+                            elements.push_back(Element::Raw(vec));
+                        }
                     }
                 }
+
+                poll_elements(
+                    &mut elements,
+                    output_writer,
+                    dispatch_fragment_request,
+                    process_fragment_response,
+                )?;
+
                 Ok(())
             },
         )?;
 
+        // Wait for any pending requests to complete
+        loop {
+            if elements.is_empty() {
+                break;
+            }
+
+            poll_elements(
+                &mut elements,
+                output_writer,
+                dispatch_fragment_request,
+                process_fragment_response,
+            )?;
+        }
+
         Ok(())
     }
+}
+
+fn build_fragment_request(mut request: Request, url: &str) -> Result<Request> {
+    let escaped_url = match quick_xml::escape::unescape(url) {
+        Ok(url) => url,
+        Err(err) => {
+            return Err(ExecutionError::InvalidRequestUrl(err.to_string()));
+        }
+    }
+    .to_string();
 
-    fn send_esi_fragment_request(
-        &self,
-        original_request: &Request,
-        url: &str,
-        request_handler: &dyn Fn(Request) -> Result<Response>,
-    ) -> Result<Response> {
-        let mut req = original_request.clone_without_body();
+    if escaped_url.starts_with('/') {
+        match Url::parse(
+            format!("{}://0.0.0.0{}", request.get_url().scheme(), escaped_url).as_str(),
```

Comment:
Looks like you could do this more simply by constructing a base `Url` however you want, and then parsing the provided string with `Url::options().base_url(Some(base)).parse(escaped_url)`. Then I believe you shouldn't have to check whether the given URL is relative.

Actually, should be base URL just be the original request's URL? Or perhaps a configuration option set when building the `Processor`?